File: app/src/main/java/com/jossy/android/mobilebenchmarkappjava/activity/BenchmarkSuiteActivity.java
package com.jossy.android.mobilebenchmarkappjava.activity;

import android.Manifest;
import android.content.Intent;
import android.os.Build;
import android.os.Bundle;
import android.os.Environment;
import android.os.Handler;
import android.os.Looper;
import android.util.Log;
import android.widget.Button;
import android.widget.ProgressBar;
import android.widget.TextView;
import android.widget.Toast;

import androidx.appcompat.app.AppCompatActivity;
import androidx.core.app.ActivityCompat;
import androidx.core.content.ContextCompat;

import com.jossy.android.mobilebenchmarkappjava.BenchmarkApplication;
import com.jossy.android.mobilebenchmarkappjava.R;
import com.jossy.android.mobilebenchmarkappjava.config.SampleConfiguration;
import com.jossy.android.mobilebenchmarkappjava.data.TestResult;

import java.io.File;
import java.io.IOException;
import java.text.SimpleDateFormat;
import java.util.Date;
import java.util.Locale;

/**
 * Aktywność orkiestrująca wykonanie suite'a testów wydajnościowych.
 * 
 * Optymalizacje dla dużych zbiorów danych (do 1M próbek):
 * 1. Buforowany zapis CSV (BufferedCsvWriter) - eliminuje wąskie gardło I/O
 * 2. Ring buffer dla podglądu w UI - ogranicza zużycie pamięci
 * 3. Asynchroniczne zapisywanie - nie blokuje pomiarów
 * 4. Checkpointy - odzyskiwanie danych przy błędach
 * 5. Konfigurowalny rozmiar próbek (100 do 1M)
 */
public class BenchmarkSuiteActivity extends AppCompatActivity {

    private static final String TAG = "BenchmarkSuiteActivity";
    private static final String BENCHMARK_TAG = "BENCHMARK"; // Tag dla synchronizacji z Pythonem
    private static final int PERMISSION_REQUEST_CODE = 123;
    private static final int ALL_TESTS = 6;
    private static final int TEST_ACTIVITY_REQUEST_CODE = 456;

    // UI Components
    private TextView currentTestInfo;
    private TextView testResults;
    private ProgressBar testProgress;
    private Button startTestsButton;
    private Button exportResultsButton;

    // Configuration - Fixed to LARGE (10,000 samples)
    private SampleConfiguration selectedConfig = SampleConfiguration.LARGE;

    // Test State
    private int currentIteration = 0;
    private int currentTestIndex = 0;
    private boolean isRunning = false;
    private long testSuiteStartTime = 0;
    private long currentIterationStartTime = 0;

    // Data Storage - Optimized for large datasets
    private final Handler handler = new Handler(Looper.getMainLooper());
    private File outputDir;
    private String sessionTimestamp;

    // Statistics
    private int totalSamplesCollected = 0;
    private int errorsEncountered = 0;

    private long appStartTime = System.currentTimeMillis();

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_benchmark_suite);

        initViews();
        setupButtons();

        if (savedInstanceState != null) {
            restoreState(savedInstanceState);
        }

        long launchTime = System.currentTimeMillis() - appStartTime;
        currentTestInfo.setText(String.format(Locale.US,
                "App Launched in: %dms\nReady to start tests.", launchTime));

        Log.d(TAG, "onCreate completed, UI initialized");
    }

    @Override
    protected void onSaveInstanceState(Bundle outState) {
        super.onSaveInstanceState(outState);
        outState.putBoolean("isRunning", isRunning);
        outState.putInt("currentTestIndex", currentTestIndex);
        outState.putInt("currentIteration", currentIteration);
        outState.putInt("totalSamplesCollected", totalSamplesCollected);
        outState.putInt("errorsEncountered", errorsEncountered);
        outState.putString("sessionTimestamp", sessionTimestamp);
        Log.i(TAG, "State saved");
    }

    private void restoreState(Bundle savedState) {
        isRunning = savedState.getBoolean("isRunning");
        currentTestIndex = savedState.getInt("currentTestIndex");
        currentIteration = savedState.getInt("currentIteration");
        totalSamplesCollected = savedState.getInt("totalSamplesCollected");
        errorsEncountered = savedState.getInt("errorsEncountered");
        sessionTimestamp = savedState.getString("sessionTimestamp");

        if (isRunning && sessionTimestamp != null) {
            try {
                restoreSession();
            } catch (IOException e) {
                Log.e(TAG, "Failed to restore session", e);
                isRunning = false;
                Toast.makeText(this, "Failed to restore session: " + e.getMessage(), Toast.LENGTH_LONG).show();
            }
        }
        updateProgressMax();
        updateProgress();
    }

    private void restoreSession() throws IOException {
        File baseDir = getExternalFilesDir(Environment.DIRECTORY_DOCUMENTS);
        outputDir = new File(baseDir, "benchmarks/" + sessionTimestamp);

        if (!outputDir.exists()) {
            // If directory is gone, we can't really resume properly, but let's try to
            // handle gracefully
            if (!outputDir.mkdirs()) {
                throw new IOException("Cannot restore output directory: " + outputDir.getAbsolutePath());
            }
        }

        Log.i(TAG, "Session restored: " + outputDir.getAbsolutePath());
    }

    private void initViews() {
        currentTestInfo = findViewById(R.id.currentTestInfo);
        testResults = findViewById(R.id.testResults);
        testProgress = findViewById(R.id.testProgress);
        startTestsButton = findViewById(R.id.startTestsButton);
        exportResultsButton = findViewById(R.id.exportResultsButton);
    }

    private void setupButtons() {
        startTestsButton.setOnClickListener(v -> {
            if (!isRunning) {
                startTestSuite();
            }
        });

        exportResultsButton.setOnClickListener(v -> exportResults());
    }

    private void updateProgressMax() {
        testProgress.setMax(selectedConfig.sampleCount * ALL_TESTS);
    }

    private void startTestSuite() {
        if (!checkPermissions()) {
            requestPermissions();
            return;
        }

        try {
            initializeSession();
        } catch (IOException e) {
            Toast.makeText(this, "Cannot initialize output: " + e.getMessage(),
                    Toast.LENGTH_SHORT).show();
            Log.e(TAG, "Failed to initialize session", e);
            return;
        }

        isRunning = true;
        currentIteration = 0;
        currentTestIndex = 0;
        totalSamplesCollected = 0;
        errorsEncountered = 0;
        testSuiteStartTime = System.currentTimeMillis();

        testResults.setText("");
        startTestsButton.setText("Uruchamianie...");
        startTestsButton.setEnabled(false);
        updateProgressMax();

        Log.i(TAG, "Starting test suite: " + selectedConfig.displayName +
                " (" + selectedConfig.sampleCount + " samples per test)");

        runNextTest();
    }

    /**
     * Inicjalizuje sesję testową - tworzy katalog wyjściowy.
     */
    private void initializeSession() throws IOException {
        sessionTimestamp = new SimpleDateFormat("yyyyMMdd_HHmmss", Locale.US).format(new Date());

        File baseDir = getExternalFilesDir(Environment.DIRECTORY_DOCUMENTS);
        outputDir = new File(baseDir, "benchmarks/" + sessionTimestamp);

        if (!outputDir.exists() && !outputDir.mkdirs()) {
            throw new IOException("Cannot create output directory: " + outputDir.getAbsolutePath());
        }

        Log.i(TAG, "Session initialized: " + outputDir.getAbsolutePath());
    }

    private void runNextTest() {
        if (currentTestIndex < ALL_TESTS) {
            String testName = getTestName(currentTestIndex);
            onTestStarted(testName);
            currentIterationStartTime = System.currentTimeMillis();
            startSpecificTest(currentTestIndex);
        } else {
            onAllTestsCompleted();
        }
    }

    private String getTestName(int index) {
        return switch (index) {
            case 0 -> "UI Stress Test"; // Changed from "UI Test"
            case 1 -> "CPU Test";
            case 2 -> "RAM Test";
            case 3 -> "Image Loading Test";
            case 4 -> "API Test";
            case 5 -> "Location Test";
            default -> "Unknown Test";
        };
    }

    private void startSpecificTest(int index) {
        if (outputDir == null) {
            Log.e(TAG, "outputDir is null in startSpecificTest. Attempting to restore or aborting.");
            // Try to recover if we have timestamp (should have been restored)
            if (sessionTimestamp != null) {
                try {
                    restoreSession();
                } catch (IOException e) {
                    Log.e(TAG, "Recovery failed", e);
                    Toast.makeText(this, "Test failed: Session lost", Toast.LENGTH_LONG).show();
                    isRunning = false;
                    startTestsButton.setEnabled(true);
                    return;
                }
            } else {
                Toast.makeText(this, "Test failed: outputDir is null", Toast.LENGTH_SHORT).show();
                isRunning = false;
                startTestsButton.setEnabled(true);
                return;
            }
        }

        Intent intent;
        switch (index) {
            case 0:
                intent = new Intent(this, GPUTestActivity.class);
                intent.putExtra("session_dir", outputDir.getAbsolutePath());
                break;
            case 1:
                intent = new Intent(this, CPUTestActivity.class);
                break;
            case 2:
                intent = new Intent(this, RAMTestActivity.class);
                break;
            case 3:
                intent = new Intent(this, ImageLoadingActivity.class);
                break;
            case 4:
                intent = new Intent(this, ApiTestActivity.class);
                break;
            default:
                intent = new Intent(this, LocationTestActivity.class);
                intent.putExtra("interval", selectedConfig.samplingIntervalMs);
                break;
        }

        // Pass common simplified args
        intent.putExtra("auto_mode", true);

        // Pass CSV path for Batch Mode tests
        // (Stress Test manages its own file, but others might use this)
        String testName = getTestName(index);
        String safeName = testName.toLowerCase(Locale.US).replace(" ", "_");
        File csvFile = new File(outputDir, safeName + ".csv");
        intent.putExtra("csv_path", csvFile.getAbsolutePath());

        startActivityForResult(intent, TEST_ACTIVITY_REQUEST_CODE);
    }

    /**
     * Obsługuje zakończenie pojedynczego testu.
     * Stosuje try-catch dla odporności na błędy.
     */
    public void onTestCompleted(TestResult result) {
        // Log completion
        long duration = System.currentTimeMillis() - currentIterationStartTime;
        Log.i(TAG, "Test " + result.getTestName() + " completed in " + duration + "ms");

        if (result.isSuccess()) {
            totalSamplesCollected += selectedConfig.sampleCount;
        }

        // Advance to next test
        currentTestIndex++;

        // Continue with next test
        handler.postDelayed(this::runNextTest, 500);
    }

    public void onAllTestsCompleted() {
        isRunning = false;
        startTestsButton.setText("Start Tests");
        startTestsButton.setEnabled(true);

        long totalTime = System.currentTimeMillis() - testSuiteStartTime;
        String summary = String.format(Locale.US,
                "All tests completed!\n" +
                        "Total samples: %d\n" +
                        "Errors: %d\n" +
                        "Total time: %.2fs\n" +
                        "Output: %s",
                totalSamplesCollected,
                errorsEncountered,
                totalTime / 1000.0,
                outputDir != null ? outputDir.getAbsolutePath() : "N/A");
        currentTestInfo.setText(summary);

        Log.i(TAG, summary.replace("\n", ", "));
    }

    public void onTestStarted(String testName) {
        Log.i(BENCHMARK_TAG, "TEST_START:" + testName);
        String info = String.format(Locale.US, "Running: %s", testName);
        currentTestInfo.setText(info);
    }

    private void updateProgress() {
        // Simplified progress: just showing which test we are on
        int progress = (currentTestIndex + 1) * selectedConfig.sampleCount;
        testProgress.setProgress(progress);
    }

    private void exportResults() {
        if (outputDir == null || !outputDir.exists()) {
            Toast.makeText(this, "No results to export yet", Toast.LENGTH_SHORT).show();
            return;
        }

        File[] files = outputDir.listFiles((dir, name) -> name.endsWith(".csv"));
        int count = files != null ? files.length : 0;

        if (count > 0) {
            Toast.makeText(this,
                    String.format(Locale.US, "Results saved: %d files in %s", count, outputDir.getAbsolutePath()),
                    Toast.LENGTH_LONG).show();
        } else {
            Toast.makeText(this, "No CSV files found", Toast.LENGTH_SHORT).show();
        }
    }

    private Boolean checkPermissions() {
        return ContextCompat.checkSelfPermission(this,
                Manifest.permission.ACCESS_FINE_LOCATION) == android.content.pm.PackageManager.PERMISSION_GRANTED;
    }

    private void requestPermissions() {
        ActivityCompat.requestPermissions(this,
                new String[] { Manifest.permission.ACCESS_FINE_LOCATION },
                PERMISSION_REQUEST_CODE);
    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, Intent data) {
        super.onActivityResult(requestCode, resultCode, data);

        if (requestCode == PERMISSION_REQUEST_CODE) {
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.R) {
                if (Environment.isExternalStorageManager()) {
                    startTestSuite();
                } else {
                    Toast.makeText(this, "Permissions required to run tests", Toast.LENGTH_SHORT).show();
                }
            }
        } else if (requestCode == TEST_ACTIVITY_REQUEST_CODE) {
            if (resultCode == RESULT_OK && data != null) {
                try {
                    TestResult result = (TestResult) data.getSerializableExtra(BenchmarkApplication.RESULT);
                    if (result != null) {
                        onTestCompleted(result);
                    } else {
                        Log.e(TAG, "No TestResult in intent");
                        errorsEncountered++;
                        handler.postDelayed(this::runNextTest, 500);
                    }
                } catch (Exception e) {
                    Log.e(TAG, "Error processing activity result: " + e.getMessage(), e);
                    errorsEncountered++;
                    handler.postDelayed(this::runNextTest, 500);
                }
            } else {
                // Test failed - log and continue
                errorsEncountered++;
                Log.w(TAG, "Test activity returned with error or cancel");
                handler.postDelayed(this::runNextTest, 500);
            }
        }
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
    }
}

File: app/src/main/java/com/jossy/android/mobilebenchmarkappkotlin/activity/BenchmarkSuiteActivity.kt
package com.jossy.android.mobilebenchmarkappkotlin.activity

import android.Manifest
import android.content.Intent
import android.content.pm.PackageManager
import android.os.Build
import android.os.Bundle
import android.os.Environment
import android.os.Handler
import android.os.Looper
import android.util.Log
import android.view.View
import android.widget.*
import androidx.appcompat.app.AppCompatActivity
import androidx.core.app.ActivityCompat
import androidx.core.content.ContextCompat
import com.jossy.android.mobilebenchmarkappkotlin.BenchmarkApplication
import com.jossy.android.mobilebenchmarkappkotlin.R
import com.jossy.android.mobilebenchmarkappkotlin.config.SampleConfiguration
import com.jossy.android.mobilebenchmarkappkotlin.io.BufferedCsvWriter
import com.jossy.android.mobilebenchmarkappkotlin.metrics.SystemMetricsCollector
import com.jossy.android.mobilebenchmarkappkotlin.model.TestEntry
import com.jossy.android.mobilebenchmarkappkotlin.model.TestResult
import java.io.File
import java.text.SimpleDateFormat
import java.util.*

/**
 * Aktywność orkiestrująca wykonanie suite'a testów wydajnościowych.
 * 
 * Optymalizacje dla dużych zbiorów danych (do 1M próbek):
 * 1. Buforowany zapis CSV (BufferedCsvWriter) - eliminuje wąskie gardło I/O
 * 2. Ring buffer dla podglądu w UI - ogranicza zużycie pamięci
 * 3. Asynchroniczne zapisywanie - nie blokuje pomiarów
 * 4. Checkpointy - odzyskiwanie danych przy błędach
 * 5. SystemMetricsCollector - zbiera CPU/RAM/GPU/FPS w tle
 */
class BenchmarkSuiteActivity : AppCompatActivity() {
    
    companion object {
        private const val TAG = "BenchmarkSuiteActivity"
        private const val BENCHMARK_TAG = "BENCHMARK"
        private const val PERMISSION_REQUEST_CODE = 123
        private const val ALL_TESTS = 6
        private const val TEST_ACTIVITY_REQUEST_CODE = 456
        private const val UI_DISPLAY_BUFFER_SIZE = 100
    }

    // UI Components
    private lateinit var currentTestInfo: TextView
    private lateinit var testResults: TextView
    private lateinit var testProgress: ProgressBar
    private lateinit var startTestsButton: Button
    private lateinit var exportResultsButton: Button

    // Configuration
    private var selectedConfig = SampleConfiguration.SMALL

    // Test State
    private var currentIteration = 0
    private var currentTestIndex = 0
    private var isRunning = false
    private var testSuiteStartTime = 0L
    private var currentIterationStartTime = 0L

    // Data Storage - Optimized for large datasets
    private val csvWriters = linkedMapOf<String, BufferedCsvWriter>()
    private val uiDisplayBuffers = linkedMapOf<String, ArrayDeque<TestEntry>>()
    private val handler = Handler(Looper.getMainLooper())
    private var outputDir: File? = null
    private var sessionTimestamp: String = ""

    // Statistics
    private var totalSamplesCollected = 0
    private var errorsEncountered = 0

    // System Metrics Collector
    private var metricsCollector: SystemMetricsCollector? = null

    private val appStartTime = System.currentTimeMillis()

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_benchmark_suite)

        initViews()
        updateProgressMax()

        val launchTime = System.currentTimeMillis() - appStartTime
        currentTestInfo.text = "App Launched in: ${launchTime}ms\nReady to start tests."
    }

    private fun initViews() {
        currentTestInfo = findViewById(R.id.currentTestInfo)
        testResults = findViewById(R.id.testResults)
        testProgress = findViewById(R.id.testProgress)
        startTestsButton = findViewById(R.id.startTestsButton)
        exportResultsButton = findViewById(R.id.exportResultsButton)

        startTestsButton.setOnClickListener {
            if (!isRunning) startTestSuite()
        }

        exportResultsButton.setOnClickListener { 
            Toast.makeText(this, "Results are saved automatically during tests", Toast.LENGTH_SHORT).show()
        }
    }

    private fun updateProgressMax() {
        testProgress.max = selectedConfig.sampleCount * ALL_TESTS
        testProgress.progress = 0
    }

    private fun startTestSuite() {
        if (checkPermissions()) {
            try {
                initializeSession()
                isRunning = true
                currentIteration = 0
                currentTestIndex = 0
                totalSamplesCollected = 0
                errorsEncountered = 0
                testSuiteStartTime = System.currentTimeMillis()

                startTestsButton.text = "Running..."
                startTestsButton.isEnabled = false
                testResults.text = ""

                runNextTest()
            } catch (e: Exception) {
                Log.e(TAG, "Failed to start test suite", e)
                Toast.makeText(this, "Failed to start: ${e.message}", Toast.LENGTH_LONG).show()
            }
        } else {
            requestPermissions()
        }
    }

    private fun initializeSession() {
        sessionTimestamp = SimpleDateFormat("yyyyMMdd_HHmmss", Locale.US).format(Date())

        val baseDir = getExternalFilesDir(Environment.DIRECTORY_DOCUMENTS)
        outputDir = File(baseDir, "benchmarks/$sessionTimestamp").apply { mkdirs() }

        // Close previous writers
        closeWriters()
        csvWriters.clear()
        uiDisplayBuffers.clear()

        // Start system metrics collector
        metricsCollector?.stop()
        metricsCollector = SystemMetricsCollector(this, outputDir!!, sessionTimestamp)
        metricsCollector?.start()

        // Initialize writers for each test - DELEGATED TO ACTIVITIES
        // We only prepare the directory here
        
        // uiDisplayBuffers for simple summary if needed
        for (i in 0 until ALL_TESTS) {
            val testName = getTestName(i)
            uiDisplayBuffers[testName] = ArrayDeque(UI_DISPLAY_BUFFER_SIZE)
        }

        Log.i(TAG, "Session initialized: ${outputDir?.absolutePath}")
    }

    private fun runNextTest() {
        if (currentTestIndex < ALL_TESTS) {
            val testName = getTestName(currentTestIndex)
            onTestStarted(testName)
            startSpecificTest(currentTestIndex)
        } else {
            onAllTestsCompleted()
        }
    }

    private fun getTestName(index: Int): String = when (index) {
        0 -> "UI Stress Test"
        1 -> "CPU Test"
        2 -> "RAM Test"
        3 -> "Image Loading Test"
        4 -> "API Test"
        5 -> "Location Test"
        else -> "Unknown Test"
    }

    private fun startSpecificTest(index: Int) {
        // Ensure output dir exists
        if (outputDir == null) {
            val baseDir = getExternalFilesDir(Environment.DIRECTORY_DOCUMENTS)
            outputDir = File(baseDir, "benchmarks/$sessionTimestamp").apply { mkdirs() }
        }

        val testName = getTestName(index)
        val safeName = testName.lowercase(Locale.US).replace(" ", "_")
        val csvPath = File(outputDir, "$safeName.csv").absolutePath

        val intent = when (index) {
            0 -> Intent(this, UITestActivity::class.java)
            1 -> Intent(this, CPUTestActivity::class.java)
            2 -> Intent(this, RAMTestActivity::class.java)
            3 -> Intent(this, ImageLoadingActivity::class.java)
            4 -> Intent(this, ApiTestActivity::class.java)
            5 -> Intent(this, LocationTestActivity::class.java)
            else -> return
        }
        
        // Pass configuration to the test activity
        intent.putExtra("sample_count", selectedConfig.sampleCount)
        intent.putExtra("output_file", csvPath)
        intent.putExtra("session_id", sessionTimestamp)
        
        // Specific extras
        if (index == 1) { // CPU Test
            intent.putExtra("cpu_iterations", selectedConfig.cpuIterationsPerThread)
        }

        intent.putExtra("auto_mode", true)
        startActivityForResult(intent, TEST_ACTIVITY_REQUEST_CODE)
    }

    private fun onTestCompleted(result: TestResult) {
        try {
            val iterationEndTime = System.currentTimeMillis()
            val cumulativeTime = iterationEndTime - testSuiteStartTime

            // Log completion
            Log.i(BENCHMARK_TAG, "TEST_END:${result.testName} Duration:${result.executionTime}ms")

            // Update UI with summary
            val summaryEntry = TestEntry(
                iteration = totalSamplesCollected, 
                result = result,
                intervalStartMs = currentIterationStartTime,
                intervalDurationMs = result.executionTime,
                cumulativeTimeMs = cumulativeTime
            )
            
            uiDisplayBuffers[result.testName]?.add(summaryEntry)

            totalSamplesCollected += selectedConfig.sampleCount // Approximate
            currentTestIndex++
            updateProgress()
            updateCsvDisplay()

        } catch (e: Exception) {
            errorsEncountered++
            Log.e(TAG, "Error processing result: ${e.message}", e)
        }

        // Continue with next test
        handler.postDelayed({ runNextTest() }, selectedConfig.testDelayMs)
    }

    private fun onAllTestsCompleted() {
        isRunning = false
        startTestsButton.text = "Start Tests"
        startTestsButton.isEnabled = true

        closeWriters()
        metricsCollector?.stop()

        val totalTime = System.currentTimeMillis() - testSuiteStartTime
        val summary = String.format(Locale.US,
            "All tests completed!\n" +
                    "Total samples: %d\n" +
                    "Errors: %d\n" +
                    "Total time: %.2fs\n" +
                    "Output: %s",
            totalSamplesCollected,
            errorsEncountered,
            totalTime / 1000.0,
            outputDir?.absolutePath ?: "N/A")
        currentTestInfo.text = summary

        calculateAndDisplayAverages()
    }

    private fun onTestStarted(testName: String) {
        // Sync metrics collector
        metricsCollector?.setCurrentTest(testName, currentIteration)

        // Log markers
        Log.i(BENCHMARK_TAG, "TEST_START:$testName")
        currentTestInfo.text = "Running: $testName\n" +
                "Target samples: ${selectedConfig.sampleCount}\n"
    }

    private fun updateProgress() {
        val progress = (currentTestIndex * selectedConfig.sampleCount) + currentIteration
        testProgress.progress = progress
    }

    private fun updateCsvDisplay() {
        val sb = StringBuilder()
        uiDisplayBuffers.forEach { (testName, entries) ->
            if (entries.isNotEmpty()) {
                sb.appendLine("# $testName (last ${entries.size})")
                entries.toList().takeLast(5).forEach { e ->
                    sb.appendLine("  [${e.iteration}] ${e.result.executionTime}ms")
                }
            }
        }
        testResults.text = sb.toString()
    }

    private fun calculateAndDisplayAverages() {
        val averages = StringBuilder("\nAverages:\n")
        uiDisplayBuffers.forEach { (testName, entries) ->
            if (entries.isNotEmpty()) {
                val avg = entries.map { it.result.executionTime }.average()
                averages.append("$testName: ${"%.2f".format(Locale.US, avg)}ms\n")
            }
        }
        testResults.append(averages.toString())
    }

    private fun closeWriters() {
        csvWriters.values.forEach { writer ->
            try {
                writer.close()
            } catch (e: Exception) {
                Log.e(TAG, "Error closing writer: ${e.message}")
            }
        }
    }

    private fun saveAllCheckpoints() {
        csvWriters.values.forEach { writer ->
            try {
                writer.flush()
            } catch (e: Exception) {
                Log.e(TAG, "Checkpoint save error: ${e.message}")
            }
        }
    }

    private fun checkPermissions(): Boolean =
        ContextCompat.checkSelfPermission(this, Manifest.permission.ACCESS_FINE_LOCATION) == PackageManager.PERMISSION_GRANTED

    private fun requestPermissions() {
        ActivityCompat.requestPermissions(this, arrayOf(Manifest.permission.ACCESS_FINE_LOCATION), PERMISSION_REQUEST_CODE)
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)
        when (requestCode) {
            PERMISSION_REQUEST_CODE -> {
                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.R) {
                    if (Environment.isExternalStorageManager()) {
                        startTestSuite()
                    } else {
                        Toast.makeText(this, "Permissions required", Toast.LENGTH_SHORT).show()
                    }
                }
            }
            TEST_ACTIVITY_REQUEST_CODE -> {
                if (resultCode == RESULT_OK) {
                    @Suppress("DEPRECATION")
                    val result = data?.getSerializableExtra(BenchmarkApplication.RESULT) as? TestResult
                    if (result != null) {
                        onTestCompleted(result)
                    } else {
                        Log.e(TAG, "No TestResult received")
                        errorsEncountered++
                        handler.postDelayed({ runNextTest() }, 100)
                    }
                }
            }
        }
    }

    override fun onDestroy() {
        super.onDestroy()
        metricsCollector?.stop()
        closeWriters()
    }
}
